content/courses/intro-to-embedded-sys-prog/chapters/interacting_with_devices.rst

```diff
@@ -3,152 +3,1633 @@ Interacting with Devices
+implement assigning a value to it. The update to the variable has only
+an internal effect. With warnings enabled we'll receive notice from
+the compiler, but they're just warnings.
+
+However, writing to the file is an external effect because the file
+persists beyond the end of the program's execution. The optimizer (when
+enabled) would be free to remove any access to the variable
+:ada:`Silly`, but not the write to the file.
+
+We can make the compiler recognize that a software object is part of an
+external effect by applying the aspect :ada:`Volatile`. (Aspect
+:ada:`Atomic` is pertinent too. More in a moment.) As a result, the
+compiler will generate load or store instructions for every read or
+update to the object that occurs in the source code. Furthermore, it
+cannot generate any additional loads or stores to that variable, and it
+cannot reorder loads or stores from their order in the source code.
+"What You See Is What You Get" in other words.
+
+.. code-block:: ada
+
+   with Ada.Text_IO;  use Ada.Text_IO;
+
+   procedure Demo is
+      Output : File_Type;
+      Silly  : Integer with Volatile;
+   begin
+      Silly := 0;
+      Create (Output, Out_File, "output.txt");
+      Put (Output, "42");
+      Close (Output);
+   end Demo;
+
+If we compile the above, even with warnings enabled we won't get any
+because the compiler is now required to generate the assignment for
+:ada:`Silly`.
+
+The variable :ada:`Silly` is not even a memory-mapped object, but
+remember that we said these aspects are important to the tasking context
+too, for shared variables. We're ignoring that context in this course.
+
+Perhaps the issue is that you want to write some Ada code that interacts
+with a memory-mapped device, and you want to have exactly the load and
+store instructions generated that match those of the Ada code. The
+memory-mapped device will be volatile, in other words, but some other
+variables might need to be volatile too.
+
+For example, :doc:`elsewhere <./multi_language_development>`
+we said that the best way to achieve exact assembly instruction
+sequences is the use of machine-code inserts (MCI). That's true, but for
+the moment let's say we want to write it in Ada without the MCIs. Our
+earlier example was the memory-mapped GPIO ports on Arm microcontrollers
+produced by ST Microelectronics. Specifically, these ports have a "lock"
+per GPIO pin that allows the developer to configure the pin and then
+lock it so that no other configuration can accidentally change the
+configuration of that pin. Doing so requires an exact sequence of loads
+and stores. If we wrote this in Ada it would look like this:
+
+.. code-block:: ada
+
+   procedure Lock (Port : in out GPIO_Port;  Pin : GPIO_Pin) is
+      Temp : Word with Volatile;
+   begin
+      --  set the lock control bit and the pin bit, clear the others
+      Temp := LCCK or Pin'Enum_Rep;
+      --  write the lock and pin bits
```

Comment:
Minor editorial suggestion: an empty line before each comment could improve readability a bit.

Reply:
Sometimes I agree with that, sometimes not so much. I think it is overdone in the AdaCore style, for example.

Reply:
I was actually just referring to the readability of this specific code example. Here, I think an empty line could look better.

Reply:
OK